### PR TITLE
[workspace] Better localstorage keys #904

### PR DESCRIPTION
**Pull request by marechal-p**
Adds the current `window.location.pathname` in the key used to store data into the local storage.

This fixes an issue where multiple theia behind a proxy would share the same local storage domain.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

### CODE REVIEW

**Comment by marechal-p** [comment]
This PR should be simplier and match more use cases (more cloud related).

File: packages/workspace/src/node/default-workspace-server.ts [hunk 1]
```diff
@@ -20,15 +20,19 @@ import { WorkspaceServer } from "../common";
...
 
     workspaceRoot = new Deferred<string | undefined>();
+    domain?: string;
 
     configure(conf: yargs.Argv): void {
         conf.usage("$0 [workspace-directory] [options]");
         conf.option('root-dir', {
             description: 'DEPRECATED: Sets the workspace directory.',
         });
+        conf.option('domain', { description: 'The domain id-like to represent', type: 'string' });
```
Comment (marcdumais-work) [inline packages/workspace/src/node/default-workspace-server.ts]: The description looks strange. Maybe reword?

File: packages/workspace/src/browser/workspace-service.ts [hunk 2]
```diff
@@ -54,6 +54,10 @@ export class WorkspaceService implements FrontendApplicationContribution {
         this.deferredRoot.resolve(this._root);
     }
 
+    async getDomain(): Promise<string | undefined> {
```
Comment (marcdumais-work) [inline packages/workspace/src/browser/workspace-service.ts]: Just curious: why did you use the name "domain", for this (I mean in general, not just on this line)?

**Comment by marcdumais-work** [comment]
> When this domain is not specified, Theia will use window.location.pathname.

If I understand correctly, the current behavior is unchanged, unless the ```--domain``` argument is used?

File: packages/workspace/src/common/workspace-protocol.ts [hunk 3]
```diff
@@ -13,6 +13,11 @@ export const workspacePath = '/services/workspace';
 export const WorkspaceServer = Symbol('WorkspaceServer');
 export interface WorkspaceServer {
 
+    /**
+     * Returns the random identifier to the running theia instance
+     */
+    getDomain(): Promise<string | undefined>;
```
Comment (svenefftinge) [inline packages/workspace/src/common/workspace-protocol.ts]: There is no need to ask the server for a domain. As we are already using a domain to connect to it.
Why not `window.location`?
Also, the Workspace doesn't have anything to do with domains.

File: packages/workspace/src/browser/workspace-storage-service.ts [hunk 4]
```diff
@@ -21,18 +22,25 @@ export class WorkspaceStorageService implements StorageService {
...
+    constructor() {
         this.storageService = new LocalStorageService(this.logger);
     }
 
+    @postConstruct()
+    protected async init(): Promise<void> {
+        const domain = await this.workspaceService.getDomain();
+        const statFile = await this.workspaceService.root;
+        const workspace = statFile ? new URI(statFile.uri).path : '_global_';
+        this.prefix = `${domain ? domain : window.location.pathname}:${workspace}`;
```
Comment (svenefftinge) [inline packages/workspace/src/browser/workspace-storage-service.ts]: You again add the pathname, only, and renamed the previous workspace id to domain? 
Can't you please just solve your issue in an extension that is specific for your workspace server?

Comment (marechal-p) [reply packages/workspace/src/browser/workspace-storage-service.ts]: Would just adding `window.location.pathname` be enough ?

File: packages/workspace/src/common/workspace-protocol.ts (hunk 3 above)
Comment (marechal-p) [reply packages/workspace/src/common/workspace-protocol.ts]: `window.location` includes everything, such as the query string (`?a=b`).
This is the reason why electron was broken before, because it adds `?port=random-number`.
You shouldn't need more than `pathname` because the local storage already segregates based on the domain/subdomain/port.

File: packages/workspace/src/browser/workspace-storage-service.ts (hunk 4 above)
Comment (svenefftinge) [reply packages/workspace/src/browser/workspace-storage-service.ts]: That would be fine with me.

File: packages/workspace/src/browser/workspace-storage-service.ts [hunk 5]
```diff
@@ -21,18 +22,24 @@ export class WorkspaceStorageService implements StorageService {
...
+    constructor() {
         this.storageService = new LocalStorageService(this.logger);
     }
 
+    @postConstruct()
+    protected async init(): Promise<void> {
+        const statFile = await this.workspaceService.root;
+        const workspace = statFile ? new URI(statFile.uri).path : '_global_';
+        this.prefix = `${window.location.pathname}:${workspace}`;
+        this.initialized = Promise.resolve();
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-storage-service.ts]: It is not the same as before and can fail at runtime with `undefined` error.

Comment (marechal-p) [reply packages/workspace/src/browser/workspace-storage-service.ts]: because `window` can be undefined ?

**Comment by mofux** [comment]
I just came across this issue in another app I'm working on and wanted to make you aware: everything you save to localstorage will be loaded into memory once your page loads (automatically) and will remain there without ever getting garbage collected. The reason for this is the synchronous nature of the localstorage api I guess. It certainly depends on your use-case and how much your really store into localstorage, but I'd advise to consider moving to a different storage backend, like indexedDB. Since your StorageService already offers async read, write and init interfaces it shouldn't be to hard to change 😅

File: packages/workspace/src/browser/workspace-storage-service.ts (hunk 5 above)
Comment (akosyakov) [reply packages/workspace/src/browser/workspace-storage-service.ts]: `initialized` is not assigned and has `undefined` value at https://github.com/theia-ide/theia/pull/1748/files#diff-ae502ec06cf26d830ece3e21e6ec038eR22

It gets assigned async only after a workspace root is resolved if in the meantime someone tries to access it he will get an error.

**Comment by akosyakov** [comment]
@mofux this remark is out of the scope of this PR, you are better to open an issue

File: packages/workspace/src/browser/workspace-storage-service.ts (hunk 5 above)
Comment (marechal-p) [reply packages/workspace/src/browser/workspace-storage-service.ts]: Good call, going to fix this, thank you.

**Comment by marcdumais-work** [comment]
@marechal-p I think this PR is ready to merge?

**Comment by marechal-p** [comment]
Moved the changes from the workspace to theia's core, because its more instance specific than workspace specific.

**Comment by marcdumais-work** [comment]
>Moved the changes from the workspace to theia's core, because its more instance specific than workspace specific.

I do not see this change reflected in this PR?

**Comment by marechal-p** [comment]
The modification is basically just this: https://github.com/theia-ide/theia/pull/1748/files#diff-daf3e3113203772304f8f0a5f6aafd7dR68

The rest is refactoring that I did while iterating over this PR (first using an ID in a file, then using `window.location.pathname` in `@theia/workspace` and finally moving it to `@theia/core`).

**Comment by marechal-p** [comment]
@svenefftinge last confirmation before I rebase and merge ?

File: packages/workspace/src/browser/workspace-storage-service.ts [hunk 6]
```diff
@@ -18,36 +20,40 @@ import { LocalStorageService } from '@theia/core/lib/browser/storage-service';
...
+
+    constructor() {
         this.storageService = new LocalStorageService(this.logger);
+        this.initialized = new Deferred<void>();
+    }
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        const statFile = await this.workspaceService.root;
+        this.prefix = statFile ? new URI(statFile.uri).path.toString() : '_global_';
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-storage-service.ts]: `new URI(statFile.uri).path.toString()` is different from `stat.uri`, is it on purpose?

Could you rollback changes which are out of scope of this PR that we don't break anything?

Comment (marechal-p) [reply packages/workspace/src/browser/workspace-storage-service.ts]:
```ts
> uri.toString()
'file:///etc/test/'
> uri.path.toString()
'/etc/test/'
>
```
It was to remove the scheme, but yes I can.

**Comment by marechal-p** [comment]
@akosyakov changes made.

**Comment by marcdumais-work** [comment]
@marechal-p code LGTM. Will test with a workspace server, before approving.
